### ISSUE TITLE
Register single by KClass argument

### ISSUE THREAD

**Issue by pauliusrum** [InsertKoinIO/koin#475]
**Is your feature request related to a problem? Please describe.**

Hi,

I'm using koin in a ktor project. In that project we have data source classes (`UsersDataSource`, `PermissionsDataSource`). These classes are used for CRUD operations.
Frontend client makes a HTTP request in which it describes with which data source it wants to communicate. 
We use classgraph library to find all data source classes (they implement the same interface). Classgraph returns a `List<KClass<*>>` which we want to register as `singles`.
Currently there is no `single` method that takes `KClass` argument instead of `reified T` generic parameter. Also there should be `create` method that takes `KClass` instead of `reified T` generic parameter. 
These data source classes are also used in internal services so we want to be able to inject them using type parameters instead of `named`.

**Describe the solution you'd like**

The solution would be to add a `single` and `create` methods that take `KClass` (or java `Class` if interoperability is important) instead of `reified T` generic parameter and register the definition by the given `KClass`.

```Kotlin
package org.koin.core.module

fun <T : Any> single(
    type: KClass<T>,
    qualifier: Qualifier? = null,
    createdAtStart: Boolean = false,
    override: Boolean = false,
    definition: Definition<T>
): BeanDefinition<T> {
    // ...
}
```

```Kotlin
package org.koin.experimental.builder

inline fun <reified T : Any> create(context: Scope): T = create(T::class, context)

fun <T : Any> create(type: KClass<T>, context: Scope): T {
    lateinit var instance: T

    val ctor = type.getFirstJavaConstructor()
    val args = getArguments(ctor, context)
    instance = ctor.makeInstance(args)

    return instance
}
```

**Comment by arnaudgiuliani** [comment]
Is available in koin-core-ext. You can do `single<MyClass>()`

**Comment by pauliusrum** [comment]
https://github.com/InsertKoinIO/koin/tree/master/koin-projects/koin-core-ext/src/main/kotlin/org/koin/experimental/builder

There is only one function that takes `reified T`. I can not use it because the type is only known at runtime not compilation. We load these classes at runtime so we only have KClass<*>. Am I missing something?

**Comment by arnaudgiuliani** [comment]
Ok I see. More a way to provide a single with KClass.

**Comment by tynn** [comment]
Have you considered using the `Qualifier`, which is a parameter already, to achieve your goal?

From an implementation perspective it might be wise to have all DSLs delegate to a non-inline function to reduce inlined code and hide implementation details better.

**Comment by pauliusrum** [comment]
> Have you considered using the Qualifier, which is a parameter already, to achieve your goal?

> These data source classes are also used in internal services so we want to be able to inject them using type parameters instead of named.

Yes we have considered `Qualifiers` but we would like to avoid `named` services since one would have to use `named` everywhere which would not be convenient.

**Comment by pauliusrum** [comment]
I upgraded from `2.0.1` to `2.1.0`. I had to make some changes to the code:

```kotlin
private fun Module.single(type: KClass<*>, createdAtStart: Boolean = false, override: Boolean = false) {
    rootScope.save(
        BeanDefinition(
            scopeDefinition = rootScope,
            primaryType = type,
            qualifier = null,
            definition = {
                val constructor = type.java.constructors.firstOrNull() ?: error("No constructor found for class '${type.java}'")
                val args = getArguments(constructor, this)

                constructor.newInstance(*args)
            },
            kind = Kind.Single,
            options = makeOptions(override, createdAtStart),
            secondaryTypes = emptyList()
        )
    )
}

// These functions were taken from 2.0.1 since https://github.com/InsertKoinIO/koin/issues/475 is not yet solved in 2.1.0
private fun getArguments(ctor: Constructor<*>, context: Scope) = ctor.parameterTypes.map { context.getWithDefault(it.kotlin) }.toTypedArray()

private fun <T : Any> Scope.getWithDefault(clazz: KClass<T>): T = get(clazz, null, null) ?: error("Koin can't be null in scope context")
```

If all of these public koin API's are stable then we can close this issue.

**Comment by pauliusrum** [comment]
I tried to update to 2.2.0-rc-1 version but the `rootScope` became `internal`. 
Are there any workarounds? 
As I understand you are not planning to add this functionality in 2.2.0?
